tianshou/data/buffer.py

```diff
@@ -520,13 +437,215 @@ def update_weight(
+            sample_num = np.bincount(buffer_idx, minlength=self.buffer_num)
+            # avoid batch_size > 0 and sample_num == 0 -> get child's all data
+            sample_num[sample_num == 0] = -1
+
+        return np.concatenate([
+            buf.sample_index(bsz) + buf.offset
+            for buf, bsz in zip(self.buffers, sample_num)
+        ])
+
+
+class CachedReplayBuffer(ReplayBuffers):
+    """CachedReplayBuffer contains a ReplayBuffers with given size and n \
+    cached buffers, cached_buffer_num * ReplayBuffer(size=max_episode_length).
+
+    The memory layout is: ``| main_buffer | cached_buffer[0] | cached_buffer[1]
+    | ... | cached_buffer[cached_buffer_num - 1]``.
+
+    The data is first stored in cached buffers. When the episode is
+    terminated, the data will move to the main buffer and the corresponding
+    cached buffer will be reset.
+
+    :param int size: the size of main buffer.
+    :param int cached_buffer_num: number of ReplayBuffer needs to be created
+        for cached buffer.
+    :param int max_episode_length: the maximum length of one episode, used in
+        each cached buffer's maxsize.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.data.ReplayBuffer` or
+        :class:`~tianshou.data.ReplayBuffers` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        size: int,
+        cached_buffer_num: int,
+        max_episode_length: int,
+        **kwargs: Any,
+    ) -> None:
+        assert cached_buffer_num > 0 and max_episode_length > 0
+        main_buffer = ReplayBuffer(size, **kwargs)
+        buffers = [main_buffer] + [ReplayBuffer(max_episode_length, **kwargs)
+                                   for _ in range(cached_buffer_num)]
+        super().__init__(buffer_list=buffers, **kwargs)
+
+    def add(  # type: ignore
+        self,
+        obs: Any,
+        act: Any,
+        rew: np.ndarray,
+        done: np.ndarray,
+        obs_next: Any = Batch(),
+        info: Optional[Batch] = Batch(),
+        policy: Optional[Batch] = Batch(),
+        cached_buffer_ids: Optional[Union[np.ndarray, List[int]]] = None,
+        **kwargs: Any,
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """Add a batch of data into CachedReplayBuffer.
+
+        Each of the data's length (first dimension) must equal to the length of
+        cached_buffer_ids. By default the cached_buffer_ids is [0, 1, ...,
+        cached_buffer_num - 1].
+
+        Return the array of episode_length and episode_reward with shape
+        (len(cached_buffer_ids), ...), where (episode_length[i],
+        episode_reward[i]) refers to the cached_buffer_ids[i]th cached buffer's
+        corresponding episode result.
+        """
+        if cached_buffer_ids is None:
+            cached_buffer_ids = np.arange(self.buffer_num - 1)
+        # in self.buffers, the first buffer is main_buffer
+        buffer_ids = np.asarray(cached_buffer_ids) + 1
```

Comment:
suggest using api we just talk about to replace '+1'

Reply:
No because self.cached_buffers is a list instead of `ReplayBuffers`